addons/mass_mailing_sms/models/sms_tracker.py

```diff
@@ -0,0 +1,79 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class SmsTracker(models.Model):
+    _inherit = "sms.tracker"
+
+    SMS_STATE_TO_TRACE_STATUS = {
+        'error': 'error',
+        'process': 'process',
+        'outgoing': 'outgoing',
+        'canceled': 'cancel',
+        'sent': 'pending',
+        'delivered': 'sent',
+    }
+
+    mailing_trace_id = fields.Many2one('mailing.trace', ondelete='cascade')
+
+    def _action_update_from_provider_error(self, provider_error):
+        status, failure_type, failure_reason = super()._action_update_from_provider_error(provider_error)
+        self._update_sms_traces(status or 'error', failure_type=failure_type, failure_reason=failure_reason)
+        return status, failure_type, failure_reason
+
+    def _action_update_from_sms_state(self, sms_state, failure_type=False):
+        super()._action_update_from_sms_state(sms_state, failure_type=failure_type)
+        trace_status = self.SMS_STATE_TO_TRACE_STATUS[sms_state]
+        traces = self._update_sms_traces(trace_status, failure_type=failure_type)
+        self._update_sms_mailings(trace_status, traces)
+
+    def _update_sms_traces(self, trace_status, failure_type=False, failure_reason=False):
+        if not self.mailing_trace_id:  # avoid a search below
+            return self.env['mailing.trace']
+        # See _update_sms_notifications
+        statuses_to_ignore = {
+            'cancel': ['cancel', 'process', 'pending', 'sent'],
+            'outgoing': ['outgoing', 'process', 'pending', 'sent'],
+            'process': ['process', 'pending', 'sent'],
+            'pending': ['pending', 'sent'],
+            'bounce': ['bounce'],
+            'sent': ['sent'],
+            'error': ['error'],
+        }[trace_status]
+        traces_sudo = self.env['mailing.trace'].search([
+            ('id', 'in', self.mailing_trace_id.ids),
+            ('trace_status', 'not in', statuses_to_ignore)
+        ]).sudo()
+        if traces_sudo:
+            traces_values = {
+                'trace_status': trace_status,
+                'failure_type': failure_type,
+                'failure_reason': failure_reason,
+                **{'sent_datetime': fields.Datetime.now() if trace_status == 'pending' else {}},
+            }
+            traces_sudo.write(traces_values)
+        return traces_sudo.sudo(False)
+
+    def _update_sms_mailings(self, trace_status, traces):
+        traces.flush_recordset(['trace_status'])
+
+        if trace_status == 'process':
+            traces.sudo().mass_mailing_id.write({'state': 'sending'})
```

Comment:
Same for sudo: as tracker is a technical model, we will probably call it as sudo ... and usage of sudo is a bit unclear to me, you flush, then you sudo. Do you have failing tests ? Otherwise would try to simplify (globally) usage of sudo in this PR, if we consider trackers are technical we can consider they are already browsed as admin / su.